### PR TITLE
INTMDB-462-teams module implementation

### PR DESCRIPTION
**Pull request by rdharmedran**
## Description

Teams resource implementation for AWS CFN
Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: cfn-resources/teams/cmd/resource/resource.go [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup() // logger setup
+
+	_, _ = logger.Debugf("Create Team - Request:%+v", currentModel)
+
+	// Validate required fields in the request
+	event, client, err := ValidateRequest(CreateRequiredFields, currentModel)
+	if err != nil {
+		return event, err
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/teams/cmd/resource/resource.go:31]: `ValidateRequest` should only validate the request. it should not create the client and the event as well

```suggestion
	err := ValidateRequest(CreateRequiredFields, currentModel); err !=nill{
		return nill, err
	}
```

File: cfn-resources/teams/cmd/resource/resource.go [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+}
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup() // logger setup
+
+	_, _ = logger.Debugf("Read Team - Request: %+v", currentModel)
+	// Validate required fields in the request
+	event, client, err := ValidateRequest(ReadRequiredFields, currentModel)
+	if err != nil {
+		return event, err
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/teams/cmd/resource/resource.go:62]: same

File: cfn-resources/teams/cmd/resource/resource.go [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+}
+func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup() // logger setup
+
+	_, _ = logger.Debugf("Update Team Request :%+v", currentModel)
+	// Validate required fields in the request
+	event, client, err := ValidateRequest(ReadRequiredFields, currentModel)
+	if err != nil {
+		return event, err
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/teams/cmd/resource/resource.go:142]: same

File: cfn-resources/teams/cmd/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup() // logger setup
+
+	_, _ = logger.Debugf("Delete Team  Request() :%+v", currentModel)
+
+	// Validate required fields in the request
+	event, client, err := ValidateRequest(CreateRequiredFields, currentModel)
+	if err != nil {
+		return event, err
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/teams/cmd/resource/resource.go:271]: same

File: cfn-resources/teams/cmd/resource/resource.go [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+					newUsers = append(newUsers, user.ID)
+				}
+			}
+		}
+		_, _ = logger.Debugf("team -Add user starts \n")
+		_, _, err = client.Teams.AddUsersToTeam(context.Background(), *currentModel.OrgId, *currentModel.TeamId, newUsers)
+		if err != nil {
+			_, _ = logger.Debugf("team -Add user response status (%d)", res.StatusCode)
+		}
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/teams/cmd/resource/resource.go:206]: this piece of code has a lot of nested if. Could you refactor it? moving some parts to specific functions would improve readability

File: cfn-resources/util/constants/constants.go [hunk 6]
```diff
@@ -50,4 +50,6 @@ const (
 	ClusterName                      = "ClusterName"
 	ErrorCreateCloudBackupRestoreJob = "Error - Create Cloud Backup Restore snapshot for Snapshot(%s)- Details: %+v"
 	ErrorReadCloudBackUpRestoreJob   = "Error - Read Restore Job with id(%s)"
+
+	TeamID = "TeamId"
```
Comment (andreaangiolillo) [inline cfn-resources/util/constants/constants.go:54]:
```suggestion
	TeamID = "TeamId"
```

**Comment by rdharmedran** [comment]
@andreaangiolillo  we have fixed all review comment.Please check

**Review by andreaangiolillo — APPROVED** [review]
LGTM

**Review by Zuhairahmed — CHANGES_REQUESTED** [review]
Thanks @rdharmedran. Since this one is missing a few key functions as below that are not in go-client can you: 1) create separate PR building out those functions and 2) build in that added functionality into this CloudFormation resource PR. Let me know if you need to discuss 

GET - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams)
DELETE - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams/{teamId}](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams/%7BteamId%7D)
POST - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams)

**Comment by rdharmedran** [comment]
> Thanks @rdharmedran. Since this one is missing a few key functions as below that are not in go-client can you: 1) create separate PR building out those functions and 2) build in that added functionality into this CloudFormation resource PR. Let me know if you need to discuss
> 
> GET - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams) DELETE - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams/{teamId}](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams/%7BteamId%7D) POST - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams)

@Zuhairahmed  We have implemented these APIs in this cloud formation module.Please review

**Comment by Zuhairahmed** [comment]
> > Thanks @rdharmedran. Since this one is missing a few key functions as below that are not in go-client can you: 1) create separate PR building out those functions and 2) build in that added functionality into this CloudFormation resource PR. Let me know if you need to discuss
> > GET - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams) DELETE - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams/{teamId}](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams/%7BteamId%7D) POST - [https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/teams](https://cloud.mongodb.com/api/atlas/v1.0/groups/%7BgroupId%7D/teams)
> 
> @Zuhairahmed We have implemented these APIs in this cloud formation module.Please review

Makes sense, thanks for the update here @rdharmedran. LGTM.

**Review by Zuhairahmed — APPROVED** [review]

**Review by govardhanpagidi — APPROVED** [review]

**Pull request merged**